src/wayland/seat/keyboard.rs

```diff
@@ -164,37 +168,79 @@ pub fn serialize_pressed_keys(keys: impl Iterator<Item = Keycode>) -> Vec<u8> {
     keys.flat_map(|key| (key.raw() - 8).to_ne_bytes()).collect()
 }
 
-impl<D: SeatHandler + 'static> KeyboardTarget<D> for WlSurface {
-    fn enter(&self, seat: &Seat<D>, state: &mut D, keys: Vec<KeysymHandle<'_>>, serial: Serial) {
-        *seat.get_keyboard().unwrap().arc.last_enter.lock().unwrap() = Some(serial);
-        for_each_focused_kbds(seat, self, |kbd| {
-            kbd.enter(
-                serial.into(),
-                self,
-                serialize_pressed_keys(keys.iter().map(|h| h.raw_code())),
-            )
-        });
+#[derive(Default)]
+pub(crate) struct FocusDestroyHook(pub Cell<Option<HookId>>);
 
-        let text_input = seat.text_input();
-        let input_method = seat.input_method();
+pub(crate) fn enter_internal<D: SeatHandler + 'static>(
+    surface: &WlSurface,
+    seat: &Seat<D>,
+    state: &mut D,
+    keys: impl Iterator<Item = Keycode>,
+    serial: Serial,
+) {
+    *seat.get_keyboard().unwrap().arc.last_enter.lock().unwrap() = Some(serial);
+    let serialized_keys = serialize_pressed_keys(keys);
+    for_each_focused_kbds(seat, surface, |kbd| {
+        kbd.enter(serial.into(), surface, serialized_keys.clone())
+    });
 
-        if input_method.has_instance() {
-            input_method.deactivate_input_method(state);
+    let seat_clone = seat.clone();
+    let hook_id = add_destruction_hook::<D, _>(surface, move |_, surface| {
+        if let Some(client) = surface.client() {
+            let keyboard = seat_clone.get_keyboard().unwrap();
+            let inner = keyboard.arc.known_kbds.lock().unwrap();
+            for kbd in &*inner {
+                let Ok(kbd) = kbd.upgrade() else {
+                    continue;
+                };
+
+                if kbd.client().is_some_and(|c| c == client) {
+                    kbd.leave(serial.into(), surface);
+                }
+            }
         }
+    });
+    with_states(surface, |states| {
+        states
+            .data_map
+            .get_or_insert::<FocusDestroyHook, _>(Default::default)
+            .0
+            .set(Some(hook_id))
```

Comment:
Do we prevent calling enter multiple times with the same target? If not this could lead to dangling destruction hooks. It would be possible to either replace the hook here removing any present hook or check before creating a new one.

Reply:
Good call